### ISSUE TITLE
As a dev, I want to automate the setup of all cloudigrade infrastructure on AWS

### ISSUE THREAD

**Issue by kdelee** [cloudigrade/cloudigrade#392]
## Summary

As a dev, I want to be able to automate the creation of all necessary AWS items so I can automate spinning up separate environments for any branch I am working on.

## Acceptance Criteria

- [ ] Verify houndigrade cluster setup is automated (manual process is documented at https://github.com/cloudigrade/cloudigrade/blob/master/docs/dev-houndigrade-cluster-setup.md )
- [ ] Verify s3 and sqs setup is automated to receive CloudTrail data ( manual process is documented at https://github.com/cloudigrade/cloudigrade/blob/master/docs/dev-cloudtrail-initial-setup.rst )
- [ ] Verify the houndigrade cluster can be torn down automatically
- [ ] Verify the s3 and sqs setup can be torn down automatically

## Assumptions and Questions

- User will provide access key ID and secret access key via environment variables to the script
- Dependency on ansible is acceptable 
- Will be called from command line (not within python)
- pip install any tools (able to isolate tools in environment)
- any other configuration will be done via environment variables (keep it CI friendly)
-

**Comment by kdelee** [comment]
Necessary to fully enable #352

**Comment by kdelee** [comment]
@elyezer I worked on this some as I was blocked on finishing my integrade PR by some cloudigrade bugs.

I have a POC that uses two ansible roles, one which I forked from someone else that I have modified very little, and another which I wrote myself.

This appears to achieve everything we want, and now I think I need some input from you regarding how this should fit into the CI workflow, where this code should live, how to deal with some dependencies, etc. 

Additionally while I have confirmed that all settings *appear* to be identical to the ones that are set up via web console and I have spun up a cloudigrade using all these ansible-created items and it *appeared* happy, there could be some invisible problem that I have not identified. There is one part of the role I forked that I don't 100% understand related to some metrics collection which I suspect **could** be stale (the last update to this fork was a couple years ago) but I have not been able to observe any issues.

```
========================== Pre-reqs =========================================
# Python executable must have boto3 importable
# Must have AWS key and id in env for acct acting as cloudigrade
# Must have ssh key created on aws account and know the name
# Must have recorded what vpc, subnet, and security group you want to use are
# Must have ansible, ansible-galaxy, and ansible-playbook in your path
# Must have boto3 module importable by /usr/bin/python executable
```

```
# ======================= Install roles and create playbook ========================
ansible-galaxy install kdelee.aws_ecs_autoscale_role
ansible-galaxy install kdelee.aws_cloudtrail_bucket
cat > cloudigrade-aws-resources.yaml << 'EOL'
- hosts: localhost
  connection: local
  roles:
    - kdelee.aws_ecs_autoscale_role
    - kdelee.aws_cloudtrail_bucket
EOL
```

The following uses some pre-configured vpc, subnet, and security groups. Alternatively we could create these each time, the same author had a roles that would do that: https://github.com/daniel-rhoades/aws-security-groups-role and https://github.com/daniel-rhoades/aws-vpc-role but it seemed like overkill to me because each "cluster" only has one instance and I think there are plenty of IPs to go around in a `/16` network, but I am willing to be corrected.
```
# =============================SET UP VARIABLES================================
# ========================== Should be able to do this on CI job ========================
# Create cluster and autoscaling group using pre-configured VPC, subnet, and
# security group
# These values are for dev09, I need access to other account to get them

export AWS_NAME_PREFIX=elijah-testing-automated-setup
export HOUNDIGRADE_ECS_CLUSTER_NAME="${AWS_NAME_PREFIX}-houndigrade"
export HOUNDIGRADE_AWS_AUTOSCALING_GROUP_NAME="${HOUNDIGRADE_ECS_CLUSTER_NAME}-asg"
export HOUNDIGRADE_LAUNCH_CONFIG_NAME="${HOUNDIGRADE_ECS_CLUSTER_NAME}-lc"
export EC2_HOST_NAME="${HOUNDIGRADE_ECS_CLUSTER_NAME}-host"
export VPC_NAME="vpc-0049bdb38cae50d7c" # grab this from list of pre-configured vpc's on aws
export SUBNET_NAME="subnet-023bec14370c7bab1" # grab this from list of pre-configured vpc's on aws
export HOUNDIGRADE_AWS_AVAILABILITY_ZONE="us-east-1c" # should be ONE zone that matches where subnet is
export SECURITY_GROUP_NAME="sg-0954d400d462ac63c" # grab this from list of pre-configured security groups
export SSH_KEY_NAME="cloudigrade-key" # needs to be exisiting ssh key name on aws
export RECOMMENDED_AMI="ami-5253c32d"
export INSTANCE_TYPE="t2.micro"
```

```
# ========================== Run playbook =====================================
ansible-playbook -e ecs_cluster_name=$HOUNDIGRADE_ECS_CLUSTER_NAME -e ec2_asg_name=$HOUNDIGRADE_AWS_AUTOSCALING_GROUP_NAME -e ec2_launch_configuration_name=$HOUNDIGRADE_LAUNCH_CONFIG_NAME -e ec2_instance_type=$INSTANCE_TYPE -e ec2_ami_id=$RECOMMENDED_AMI -e ec2_instance_name=$EC2_HOST_NAME -e ec2_asg_min_size=0 -e ec2_asg_max_size=0 -e ec2_asg_desired_capacity=0 -e ec2_asg_availability_zones=$HOUNDIGRADE_AWS_AVAILABILITY_ZONE -e vpc_name=$VPC_NAME -e ec2_asg_vpc_subnets=$SUBNET_NAME -e ec2_security_groups=$SECURITY_GROUP_NAME -e key_name=$SSH_KEY_NAME -e application_name=houndigrade -e aws_prefix=$AWS_NAME_PREFIX cloudigrade-aws-resources.yaml
```

```
# ======================== Observe Results =======================================
# show autoscaling group
# note there is only one availability zone and 0 for min max and default instances
aws autoscaling describe-auto-scaling-groups --auto-scaling-group-names $HOUNDIGRADE_AWS_AUTOSCALING_GROUP_NAME

# show cluster
aws ecs list-clusters | grep $HOUNDIGRADE_ECS_CLUSTER_NAME

# show s3 bucket
aws s3 ls | grep $AWS_NAME_PREFIX
# go to web UI to show policy and notifications ... would be good to find CLI way to confirm

# show queue 
aws sqs list-queues | grep "${AWS_NAME_PREFIX}-cloudigrade-cloudtrail-sqs"
# go to web UI to show policy ... would be good to find CLI way to confirm


# change into shiftigrade and push templates
make oc-create-cloudigrade-all
oc start-build cloudigrade-api
```

```
# ========================= Clean up ==========================================
# Remove everything we made
ansible-playbook -e ecs_cluster_name=$HOUNDIGRADE_ECS_CLUSTER_NAME -e ec2_asg_name=$HOUNDIGRADE_AWS_AUTOSCALING_GROUP_NAME -e ec2_launch_configuration_name=$HOUNDIGRADE_LAUNCH_CONFIG_NAME -e ecs_state=absent -e ec2_lc_state=absent -e ec2_asg_state=absent -e bucket_state=absent -e sqs_state=absent -e aws_prefix=$AWS_NAME_PREFIX cloudigrade-aws-resources.yaml

# Show clean up
# show no cluster
aws ecs list-clusters | grep $HOUNDIGRADE_ECS_CLUSTER_NAME

# show no s3 bucket
aws s3 ls | grep $AWS_NAME_PREFIX

# show no queue 
aws sqs list-queues | grep "${AWS_NAME_PREFIX}-cloudigrade-cloudtrail-sqs"
```

**Comment by kdelee** [comment]
So top questions are
1) Is it OK to depend on pre-configured vpc, subnet and security group?
2) How to deal with my dependency on boto3 being importable ... should I assume no virtual env and have the anisble role install it to the machine? What OS are we going to run this on? Or should I assume ansible and everything is in a virtual env? I had to do this because no ansible module had a way to interact with creating bucket notifications, nor did `aws` cli have a way. Only ways were through boto or a direct API request.
3) Where should this code live and is it good to use ansible-galaxy? I like it so far but not 100% sure on the workflow of making sure roles are updated from ansible-galaxy (not sure if you have to kick off an import or if it automatically tracks master)

**Comment by kdelee** [comment]
@ironfroggy If you can try this out, I think the consensus we came to was that once it works for you as a user on the CLI, we will close this issue and pick up any further work in another issue, i.e. if we want to migrate the code to the cloudigrade namespace or want to make any further tweaks to make it more CI friendly.

**Comment by kdelee** [comment]
@ironfroggy can you confirm verification items are covered by https://github.com/cloudigrade/cloudigrade/blob/master/docs/dev-automated-aws-setup.md and close this issue?

**Comment by kdelee** [comment]
As I was going over this with @dsesami we realized that CLOUDTRAIL_EVENT_URL just got added last week to shiftigrade, and so you'd need to set that to get the whole data round trip to work with the s3 bucket notification. Made PR #408 to include this info
